go/samples/rag/main.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This program can be manually tested like so:
+//
+// In development mode (with the environment variable GENKIT_ENV="dev"):
+// Start the server listening on port 3100:
+//
+//	go run . &
+//
+// Tell it to run a flow:
+//
+//	curl -d '{"key":"/flow/simpleQaFlow/simpleQaFlow", "input":{"start": {"input":{"question": "What is the capital of UK?"}}}}' http://localhost:3100/api/runAction
+//
+// In production mode (GENKIT_ENV missing or set to "prod"):
+// Start the server listening on port 3400:
+//
+//	go run . &
+//
+// Tell it to run a flow:
+//
+//   curl -d '{"question": "What is the capital of UK?"}' http://localhost:3400/simpleQaFlow
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/firebase/genkit/go/ai"
+	"github.com/firebase/genkit/go/genkit"
+	"github.com/firebase/genkit/go/plugins/dotprompt"
+	"github.com/firebase/genkit/go/plugins/googleai"
+	"github.com/firebase/genkit/go/plugins/localvec"
+	"github.com/invopop/jsonschema"
+)
+
+const simpleQaPromptTemplate = `
+You're a helpful agent that answers the user's common questions based on the context provided.
+
+Here is the user's query: {{query}}
+
+Here is the context you should use: {{context}}
+
+Please provide the best answer you can.
+`
+
+type simpleQaInput struct {
+	Question string `json:"question"`
+}
+
+type simpleQaPromptInput struct {
+	Query   string `json:"query"`
+	Context string `json:"context"`
+}
+
+func main() {
+	apiKey := os.Getenv("GOOGLE_GENAI_API_KEY")
+	if apiKey == "" {
+		fmt.Fprintln(os.Stderr, "coffee-shop example requires setting GOOGLE_GENAI_API_KEY in the environment.")
```

Comment:
This is now the rag example, not the coffee-shop example.

Reply:
Done.